src/sync/remote_mongo_collection.cpp

```diff
@@ -18,3 +18,472 @@
+                            std::function<void(std::string, util::Optional<AppError>)> completion_block)
+{
+    if (value && !error) {
+        return completion_block(*value, error);
+    }
+    
+    return completion_block("", error);
+}
+
+static void handle_response_optional(util::Optional<AppError> error,
+                            util::Optional<std::string> value,
+                            std::function<void(Optional<std::string>, util::Optional<AppError>)> completion_block)
+{
+    if (value && !error) {
+        //response can be a http 200 and return "null" in the body
+        if (value && *value != "null") {
+            return completion_block(*value, error);
+        } else {
+            return completion_block(util::none, error);
+        }
+    }
+    
+    return completion_block(util::none, error);
+}
+
+static void handle_count_response(util::Optional<AppError> error,
+                                  util::Optional<std::string> value,
+                                  std::function<void(uint64_t, util::Optional<AppError>)> completion_block)
+{
+    if (value && !error) {
+        try {
+            auto json = nlohmann::json::parse(*value);
+            auto count_string = json.at("$numberLong").get<std::string>();
+            return completion_block(std::stoll(count_string), error);
+        } catch (const std::exception& e) {
+            return completion_block(0, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+    }
+    
+    return completion_block(0, error);
+}
+
+static void handle_delete_count_response(util::Optional<AppError> error,
+                                         util::Optional<std::string> value,
+                                         std::function<void(uint64_t, util::Optional<AppError>)> completion_block)
+{
+    if (value && !error) {
+        try {
+            auto json = nlohmann::json::parse(*value);
+            auto count_string = json.at("deletedCount").at("$numberInt").get<std::string>();
+            return completion_block(std::stoi(count_string), error);
+        } catch (const std::exception& e) {
+            return completion_block(0, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+    }
+    
+    return completion_block(0, error);
+}
+
+static void handle_update_response(util::Optional<AppError> error,
+                                   util::Optional<std::string> value,
+                                   std::function<void(RemoteMongoCollection::RemoteUpdateResult, util::Optional<AppError>)> completion_block)
+{
+    if (value && !error) {
+        
+        try {
+            auto json = nlohmann::json::parse(*value);
+            auto matched_count_string = json.at("matchedCount").at("$numberInt").get<std::string>();
+            auto matched_count = std::stoull(matched_count_string);
+            auto modified_count_string = json.at("modifiedCount").at("$numberInt").get<std::string>();
```

Comment:
it would be nice to have constants here instead of hard coding these value strings, can we reuse the constants that @jsflax added recently to support extended bson?

Reply:
For the moment we will keep this current implementation, I do agree constants would be good. In a subsequent PR we will move to using BSON completely which removed the need for the hard coded strings.